src/main/java/com/yelp/nrtsearch/server/luceneserver/SearchHandler.java

```diff
@@ -169,6 +171,11 @@ public SearchResponse handle(IndexState indexState, SearchRequest searchRequest)
             .addAllFacetResult(
                 FacetTopDocs.facetTopDocsSample(
                     hits, searchRequest.getFacetsList(), indexState, s.searcher, diagnostics));
+      } else if (isCompletionQuery(searchContext)) {
+        SuggestIndexSearcher suggestIndexSearcher =
+            new SuggestIndexSearcher(s.searcher.getIndexReader());
+        CompletionQuery completionQuery = (CompletionQuery) searchContext.getQuery();
+        hits = suggestIndexSearcher.suggest(completionQuery, searchContext.getTopHits(), true);
```

Comment:
This will first order work, but has some down sides. The search will be done single threaded, and you won't have access to the `key` or `context` fields in the `SuggestScoreDoc`.

A cleaner way would be to check the query type here https://github.com/Yelp/nrtsearch/blob/master/src/main/java/com/yelp/nrtsearch/server/luceneserver/search/SearchRequestProcessor.java#L330 and provide a new `DocCollector` implementation for suggest.

Here is a simple example https://github.com/Yelp/nrtsearch/blob/master/src/main/java/com/yelp/nrtsearch/server/luceneserver/search/collectors/HitCountCollector.java For your case, `newCollector()` will provide a new `TopSuggestDocsCollector` (See `SuggestIndexSearcher` implementation), and `reduce()` will get the `TopSuggestDocs` from each Collector and use `TopSuggestDocs.merge()` to combine them.

The `ScoreDoc` passed to `fillHitRanking()` will be of type `SuggestScoreDoc`, and can be used to put the key/context fields into the response along with the score if desired. https://github.com/Yelp/nrtsearch/blob/master/src/main/java/com/yelp/nrtsearch/server/luceneserver/search/collectors/RelevanceCollector.java#L64

Reply:
Thanks for the guidance!

It turns out there is no need to use `SuggestIndexSearcher` after all using the approach you have suggested here. I have updated the code with your suggestion, using a custom collector (with collector manager) for completion queries.